### PR TITLE
Improve value propagation handlers for integer truncation

### PR DESCRIPTION
**Pull request by jdmpapin**
Previously, the result would be unconstrained whenever it was possible for the child to take on the (sign-extension of) the min or max value of the result type.

The result would also be unconstrained whenever it was possible for the child to take on a value outside of the signed range of the result type. This was particularly unhelpful for a constant child.

Finally, the truncation node would never be constant-folded even when its result was constrained to be constant.

Now, a range constraint is produced whenever possible. For example, if the child of `i2b` is constrained to [511, 513], then the `i2b` will now be constrained to [-1, 1]. If OTOH the child is constrained to [383, 384], then the result will be unconstrained. When the child is constant, it is always possible to constrain, and the resulting constraint is also constant. In this case, the node is now folded.

In particular, this allows VP to see the result of an is-array test for a known class on 64-bit even when the flags are out of signed 32-bit range, e.g.
```
iushr
  iand
    l2i (*)
      lconst 0x8yyyyyyy
    iconst 0x10000
  iconst 16
```
Previously, failure to constrain `l2i` would stop propagation there.

----

Additionally:
- Delete VP handler `constrainNarrowToChar()`
- Stop using integer truncation VP handlers for FP conversions
- Simplify `(i2s (l2i x))` to `(l2s x)`

### CODE REVIEW

**Comment by jdmpapin** [comment]
@vijaysun-omr, would you mind reviewing?

File: compiler/optimizer/VPHandlers.cpp [hunk 1]
```diff
@@ -8321,67 +8321,195 @@ void replaceWithSmallerType(OMR::ValuePropagation *vp, TR::Node *node)
...
+      node->setIsNonNegative(true);
+
+   // Constrain the result.
+   switch (node->getDataType())
+      {
+      case TR::Int16:
+         constraint = TR::VPShortRange::create(vp, (int16_t)low, (int16_t)high);
+         break;
+
+      case TR::Int8: // byte nodes are still using Int constraint
```
Comment (vijaysun-omr) [inline compiler/optimizer/VPHandlers.cpp:8495]: This is an odd state of affairs that I wasn't aware of, i.e. byte nodes using int constraints. But I realize this was true before this change from you anyway. I did double check that `replaceByConstant` obtained the constraint for byte node/case as an integer one.

**Review by vijaysun-omr — APPROVED** [review]

**Comment by vijaysun-omr** [comment]
Jenkins build all

**Comment by vijaysun-omr** [comment]
It may be worth opening a "beginner issue" for someone to clean up that oddity with respect to byte type nodes (assuming you agree it would be better to be consistent).

**Comment by vijaysun-omr** [comment]
The PPC64LE Linux failure is https://github.com/eclipse/omr/issues/6571:

[2022-10-14T00:03:56.123Z] 30: [----------] 14 tests from Special/PPCDirectEncodingTest
[2022-10-14T00:03:56.123Z] 30: free(): invalid next size (normal)
[2022-10-14T00:03:56.123Z] 30/30 Test #30: compunittest ......................Child aborted***Exception:   0.80 sec
...
[2022-10-14T00:03:56.123Z] The following tests FAILED:
[2022-10-14T00:03:56.123Z] 	 30 - compunittest (Child aborted)
[2022-10-14T00:03:56.123Z] Errors while running CTest

**Comment by vijaysun-omr** [comment]
Since the other checks have passed, I am merging this.

**Comment by jdmpapin** [comment]
Opened #6778

**Pull request merged**